prusti-encoder/src/encoders/type/typ.rs

```diff
@@ -0,0 +1,1373 @@
+    pub snapshot: vir::Type<'vir>,
+    //pub method_refold: &'vir str,
+    pub specifics: TypeEncoderOutputRefSub<'vir>,
+}
+impl<'vir> task_encoder::OutputRefAny for TypeEncoderOutputRef<'vir> {}
+
+impl<'vir> TypeEncoderOutputRef<'vir> {
+    pub fn expect_structlike(&self) -> &TypeEncoderOutputRefSubStruct<'vir> {
+        match &self.specifics {
+            TypeEncoderOutputRefSub::StructLike(data) => data,
+            _ => panic!("expected structlike type"),
+        }
+    }
+    pub fn expect_prim(&self) -> &TypeEncoderOutputRefSubPrim<'vir> {
+        match &self.specifics {
+            TypeEncoderOutputRefSub::Primitive(prim) => prim,
+            _ => panic!("expected primitive type"),
+        }
+    }
+    pub fn expect_enumlike(&self) -> &TypeEncoderOutputRefSubEnum<'vir> {
+        match &self.specifics {
+            TypeEncoderOutputRefSub::EnumLike(enu) => enu,
+            _ => panic!("expected enumlike type"),
+        }
+    }
+    pub fn expect_variant(&self, vid: abi::VariantIdx) -> &TypeEncoderOutputRefSubStruct<'vir> {
+        match &self.specifics {
+            TypeEncoderOutputRefSub::StructLike(s) => {
+                assert_eq!(vid, abi::FIRST_VARIANT);
+                s
+            }
+            TypeEncoderOutputRefSub::EnumLike(e) => &e.variants[vid.as_usize()].0,
+            _ => panic!("expected struct or enum type"),
+        }
+    }
+    pub fn expect_variant_opt(&self, vid: Option<abi::VariantIdx>) -> &TypeEncoderOutputRefSubStruct<'vir> {
+        match &self.specifics {
+            TypeEncoderOutputRefSub::StructLike(s) => {
+                assert_eq!(vid, None);
+                s
+            }
+            TypeEncoderOutputRefSub::EnumLike(e) => &e.variants[vid.unwrap().as_usize()].0,
+            _ => panic!("expected struct or enum type"),
+        }
+    }
+
+    pub fn expr_from_u128(&self, val: u128) -> vir::Expr<'vir> {
+        match self.expect_prim().prim_type {
+            vir::TypeData::Bool => vir::with_vcx(|vcx| {
+                self.expect_prim().prim_to_snap.apply(vcx, [vcx.mk_const_expr(vir::ConstData::Bool(val != 0))])
+            }),
+            vir::TypeData::Int { .. } => vir::with_vcx(|vcx| {
+                self.expect_prim().prim_to_snap.apply(vcx, [
+                    vcx.mk_const_expr(vir::ConstData::Int(val))
+                ])
+            }),
+            k => {
+                tracing::error!(
+                    "unsupported type in expr_from_u128: {self:?} | {k:?} | ({:?})",
+                    self.snapshot
+                );
+                vir::with_vcx(|vcx| {
+                    vcx.alloc(vir::ExprData::Const(vcx.alloc(vir::ConstData::Int(val))))
+                })
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct TypeEncoderOutput<'vir> {
+    pub fields: &'vir [vir::Field<'vir>],
+    pub snapshot: vir::Domain<'vir>,
+    pub predicate: vir::Predicate<'vir>,
+    pub other_predicates: &'vir [vir::Predicate<'vir>],
```

Comment:
Why not only have the slice of predicates? Do we treat the primary `predicate` specially anywhere?